ssl/ssl_lib.c

```diff
@@ -3867,12 +3859,75 @@ SSL *SSL_dup(SSL *s)
         if ((ret->cipher_list_by_id = sk_SSL_CIPHER_dup(s->cipher_list_by_id))
             == NULL)
             goto err;
+    if (s->tls13_ciphersuites != NULL) {
+        if ((ret->tls13_ciphersuites = sk_SSL_CIPHER_dup(s->tls13_ciphersuites))
+                == NULL)
+            goto err;
+    }
+
+    if (s->cert != NULL) {
+        ret->cert = ssl_cert_dup(s->cert);
+        if (ret->cert == NULL)
+            goto err;
+    }
+
+    if (s->session != NULL) {
+        if (!SSL_SESSION_up_ref(s->session))
+            goto err;
+        ret->session = s->session;
+    }
+
+    if (s->psksession != NULL) {
+        if (!SSL_SESSION_up_ref(s->psksession))
+            goto err;
+        ret->psksession = s->psksession;
+    }
+
+    if (!CRYPTO_dup_ex_data(CRYPTO_EX_INDEX_SSL, &ret->ex_data, &s->ex_data))
+        goto err;
 
-    /* Dup the client_CA list */
     if (!dup_ca_names(&ret->ca_names, s->ca_names)
             || !dup_ca_names(&ret->client_ca_names, s->client_ca_names))
         goto err;
 
+    if (s->ext.supportedgroups != NULL) {
+        ret->ext.supportedgroups =
+            OPENSSL_memdup(s->ext.supportedgroups,
+                       s->ext.supportedgroups_len
+                       * sizeof(*s->ext.supportedgroups));
+        if (ret->ext.supportedgroups == NULL)
+            goto err;
+    }
+
+    if (s->ext.alpn != NULL) {
+        ret->ext.alpn = OPENSSL_malloc(s->ext.alpn_len);
+        if (ret->ext.alpn == NULL)
+            goto err;
+        memcpy(ret->ext.alpn, s->ext.alpn, s->ext.alpn_len);
+    }
+
+    if (!SSL_CTX_up_ref(s->session_ctx))
+        goto err;
+    ret->session_ctx = s->session_ctx;
+
+    if (s->srtp_profiles != NULL) {
+        ret->srtp_profiles = sk_SRTP_PROTECTION_PROFILE_dup(s->srtp_profiles);
+        if (ret->srtp_profiles == NULL)
+            goto err;
+    }
+
+    if (!ssl_srp_ctx_init_int(ret, &s->srp_ctx))
```

Comment:
shouldn't statement `if (!ssl_srp_ctx_init_int(ret, &s->srp_ctx))` (also the declaration of ssl_srp_ctx_init_int) be included by `#ifndef OPENSSL_NO_SRP` ?

Reply:
Yes. You're right. Fixed.